lib/rubygems/commands/uninstall_command.rb

```diff
@@ -148,21 +148,30 @@ def uninstall_all
 
   def uninstall_specific
     deplist = Gem::DependencyList.new
+    gems_to_uninstall = []
 
     get_all_gem_names_and_versions.each do |name, version|
-      requirement = Array(version || options[:version])
-      gem_specs = Gem::Specification.find_all_by_name(name, *requirement)
+      requirement = version || options[:version]
+      gems_to_uninstall << { name: name, version: requirement }
+      gem_specs = Gem::Specification.find_all_by_name(name, requirement)
+
       say("Gem '#{name}' is not installed") if gem_specs.empty?
-      gem_specs.each do |spec|
-        deplist.add spec
-      end
+      gem_specs.each { |spec| deplist.add spec }
     end
 
     deps = deplist.strongly_connected_components.flatten.reverse
 
-    deps.each do |dep|
-      options[:version] = dep.version
-      uninstall_gem(dep.name)
+    gems_to_uninstall.each do |dep|
+      if dep[:version] == Gem::Requirement.default
+        deps.map(&:name).uniq.each do |gem_name|
+          uninstall_gem(gem_name)
+        end
+      else
+        deps.each do |dep|
+          options[:version] = dep.version
+          uninstall_gem(dep.name)
+        end
+      end
```

Comment:
I think this nested loop might be creating some issues? See

### Currently

```
$ gem install rails:5.0.7.1 rails:5.1.6.1 rails:5.2.2 zeitwerk
Fetching rails-5.0.7.1.gem
Successfully installed rails-5.0.7.1
Successfully installed rails-5.1.6.1
Successfully installed rails-5.2.2
Fetching zeitwerk-1.3.1.gem
Successfully installed zeitwerk-1.3.1
4 gems installed

$ gem uninstall rails zeitwerk
Successfully uninstalled zeitwerk-1.3.1
Successfully uninstalled rails-5.0.7.1
Successfully uninstalled rails-5.1.6.1
Successfully uninstalled rails-5.2.2
```

### With this branch

```
$ RUBYOPT=-Ilib bin/gem install rails:5.0.7.1 rails:5.1.6.1 rails:5.2.2 zeitwerk
Fetching rails-5.0.7.1.gem
Successfully installed rails-5.0.7.1
Fetching rails-5.1.6.1.gem
Successfully installed rails-5.1.6.1
Fetching rails-5.2.2.gem
Successfully installed rails-5.2.2
Fetching zeitwerk-1.3.1.gem
Successfully installed zeitwerk-1.3.1
4 gems installed

$ RUBYOPT=-Ilib bin/gem uninstall rails zeitwerk
Successfully uninstalled zeitwerk-1.3.1

Select gem to uninstall:
 1. rails-5.0.7.1
 2. rails-5.1.6.1
 3. rails-5.2.2
 4. All versions
> 1
Successfully uninstalled rails-5.0.7.1

Select gem to uninstall:
 1. rails-5.1.6.1
 2. rails-5.2.2
 3. All versions
> 2
Successfully uninstalled rails-5.2.2
```

It's asking me the same question twice, which I already answered, so I don't know what to do.

Reply:
ah nice catch, would add a test for that edge case